src/warp.jl

```diff
@@ -124,8 +124,28 @@ julia> imrotate(img, π/4, Constant())
 
 See also [`warp`](@ref).
 """
-function imrotate(img::AbstractArray{T}, θ::Real, args...) where T
-    θ = floor(mod(θ,2pi)*typemax(Int16))/typemax(Int16) # periodic discretezation
-    tform = recenter(RotMatrix{2}(θ), center(img))
-    warp(img, tform, args...)
+function imrotate(img::AbstractMatrix{T}, θ::Real, args...) where T
+    # 1. discretize periodic for numerical stability to make sure
+    # imrotate(img, θ+2pi) == imrotate(img, θ)
+    # 2. typemax(Int16) is 32767, we choose 32760 to make sure the
+    # discretization result of pi/2 is exactly pi/2 (or 90°)
+    max_num_angles = 32760
+    θ = round(Int, 180*floor(mod(θ, 2pi)/pi*max_num_angles)/max_num_angles)
+    tform = recenter(RotMatrix{2}(θ/180*pi), center(img))
+    if θ == 0
+        return img
+    elseif θ == 90
+        idx = StepRange.(axes(img))
+        perm_img = PermutedDimsArray(img, (2, 1))
+        return view(perm_img, idx[2], reverse(idx[1]))
+    elseif θ == 180
+        idx = map(i->1:1:length(i), axes(img))
+        return view(img, reverse(idx[1]), reverse(idx[2]))
+    elseif θ == 270
+        idx = StepRange.(axes(img))
+        perm_img = PermutedDimsArray(img, (2, 1))
+        return view(perm_img, reverse(idx[2]), idx[1])
```

Comment:
any reason to not use some of the tools in base?

https://docs.julialang.org/en/v1/base/arrays/#Base.rot180
https://docs.julialang.org/en/v1/base/arrays/#Base.rotl90
https://docs.julialang.org/en/v1/base/arrays/#Base.rotr90

Reply:
I don't think @johnnychen94 want to reallocate the memory and relocate the elements.:smile:

Reply:
True, and there won't be type stability issues if we collect them and return an  `Array`; but that would be slower than views.

I don't have a good fix in mind for this at the moment.